File: slow-methods-succeed-but-never-match/python/match_delayed_payments.py
"""Match SOFORT, Klarna, and other delayed methods that succeeded after checkout
to the WooCommerce order they belong to. Run on a schedule. Safe to run again and again.

Guide: https://www.allanninal.dev/woocommerce/slow-methods-succeed-but-never-match/
"""
import os
import time
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("match_delayed_payments")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_HOURS = int(os.environ.get("LOOKBACK_HOURS", "72"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

# Payment method types that confirm asynchronously, sometimes hours after checkout.
DELAYED_METHODS = {"sofort", "klarna", "sepa_debit", "bancontact", "ideal"}
PAID_STATUSES = {"processing", "completed"}
CLOSED_STATUSES = {"cancelled", "refunded", "failed", "trash"}


def recent_succeeded_delayed(lookback_hours):
    """Yield Stripe PaymentIntents that are succeeded and used a delayed method."""
    since = int(time.time()) - lookback_hours * 3600
    for intent in stripe.PaymentIntent.list(limit=100, created={"gte": since}).auto_paging_iter():
        if intent.status != "succeeded":
            continue
        methods = set(intent.get("payment_method_types") or [])
        if methods & DELAYED_METHODS:
            yield intent


def get_order(order_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def find_order_by_intent(intent_id):
    """Fallback lookup for when the PaymentIntent has no order_id in metadata."""
    r = requests.get(
        f"{WOO_URL}/wp-json/wc/v3/orders",
        params={"search": intent_id, "per_page": 5},
        auth=AUTH, timeout=30,
    )
    r.raise_for_status()
    matches = r.json()
    return matches[0] if matches else None


def resolve_order(intent):
    order_id = intent.metadata.get("order_id")
    order = get_order(order_id) if order_id else None
    if order is not None:
        return order
    return find_order_by_intent(intent["id"])


def order_amount_minor(order):
    # Works for two decimal currencies. Zero decimal currencies (JPY and friends)
    # need their own rounding rule, since 50.00 is wrong for those.
    return round(float(order["total"]) * 100)


def decide(order, intent):
    """Pure decision function: no I/O, easy to unit test.

    Returns a (action, reason) tuple where action is one of:
    "fix", "skip", "mismatch", "orphan".
    """
    if intent.get("status") != "succeeded":
        return ("skip", "intent not succeeded")
    if order is None:
        return ("orphan", "order not found")
    if order["status"] in PAID_STATUSES:
        return ("skip", "order already paid")
    if order["status"] in CLOSED_STATUSES:
        return ("skip", "order already closed")
    if order.get("currency", "").lower() != intent.get("currency", "").lower():
        return ("mismatch", "currency does not match")
    if abs(order_amount_minor(order) - intent["amount_received"]) > 1:
        return ("mismatch", "amount does not match")
    return ("fix", "delayed method succeeded, order never caught up")


def mark_processing(order_id, intent):
    charge_id = intent.get("latest_charge") or intent["id"]
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}",
        json={
            "status": "processing",
            "transaction_id": charge_id,
            "meta_data": [{"key": "_stripe_intent_id", "value": intent["id"]}],
        },
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/notes",
        json={"note": f"Matched to Stripe PaymentIntent {intent['id']} ({intent.get('payment_method_types')}), "
                      f"which confirmed after checkout. Marked processing by the reconciler."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    fixed = 0
    for intent in recent_succeeded_delayed(LOOKBACK_HOURS):
        order = resolve_order(intent)
        order_id = order["id"] if order else intent.metadata.get("order_id")
        action, reason = decide(order, intent)
        if action == "orphan":
            log.warning("Intent %s has no matching order", intent.id)
            continue
        if action in ("skip", "mismatch"):
            if action == "mismatch":
                log.warning("Order %s: %s", order_id, reason)
            continue
        log.info("Order %s: %s. %s", order_id, reason, "would fix" if DRY_RUN else "fixing")
        if not DRY_RUN:
            mark_processing(order_id, intent)
        fixed += 1
    log.info("Done. %d order(s) %s.", fixed, "to fix" if DRY_RUN else "fixed")


if __name__ == "__main__":
    run()

File: slow-methods-succeed-but-never-match/node/match-delayed-payments.js
/**
 * Match SOFORT, Klarna, and other delayed methods that succeeded after checkout
 * to the WooCommerce order they belong to. Run on a schedule. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/woocommerce/slow-methods-succeed-but-never-match/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_HOURS = Number(process.env.LOOKBACK_HOURS || 72);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

// Payment method types that confirm asynchronously, sometimes hours after checkout.
const DELAYED_METHODS = new Set(["sofort", "klarna", "sepa_debit", "bancontact", "ideal"]);
const PAID_STATUSES = new Set(["processing", "completed"]);
const CLOSED_STATUSES = new Set(["cancelled", "refunded", "failed", "trash"]);

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

export async function* recentSucceededDelayed(lookbackHours) {
  const since = Math.floor(Date.now() / 1000) - lookbackHours * 3600;
  for await (const intent of stripe.paymentIntents.list({ limit: 100, created: { gte: since } })) {
    if (intent.status !== "succeeded") continue;
    const methods = intent.payment_method_types || [];
    if (methods.some((m) => DELAYED_METHODS.has(m))) yield intent;
  }
}

export async function getOrder(orderId) {
  return woo(`/orders/${orderId}`);
}

export async function findOrderByIntent(intentId) {
  const matches = await woo(`/orders?search=${encodeURIComponent(intentId)}&per_page=5`);
  return matches && matches.length ? matches[0] : null;
}

export async function resolveOrder(intent) {
  const orderId = intent.metadata.order_id;
  const order = orderId ? await getOrder(orderId) : null;
  if (order) return order;
  return findOrderByIntent(intent.id);
}

export function orderAmountMinor(order) {
  // Works for two decimal currencies. Zero decimal currencies (JPY and friends)
  // need their own rounding rule, since 50.00 is wrong for those.
  return Math.round(parseFloat(order.total) * 100);
}

/**
 * Pure decision function: no I/O, easy to unit test.
 * Returns a [action, reason] tuple where action is one of:
 * "fix", "skip", "mismatch", "orphan".
 */
export function decide(order, intent) {
  if (intent.status !== "succeeded") return ["skip", "intent not succeeded"];
  if (!order) return ["orphan", "order not found"];
  if (PAID_STATUSES.has(order.status)) return ["skip", "order already paid"];
  if (CLOSED_STATUSES.has(order.status)) return ["skip", "order already closed"];
  if ((order.currency || "").toLowerCase() !== (intent.currency || "").toLowerCase()) {
    return ["mismatch", "currency does not match"];
  }
  if (Math.abs(orderAmountMinor(order) - intent.amount_received) > 1) {
    return ["mismatch", "amount does not match"];
  }
  return ["fix", "delayed method succeeded, order never caught up"];
}

export async function markProcessing(orderId, intent) {
  const chargeId = intent.latest_charge || intent.id;
  await woo(`/orders/${orderId}`, {
    method: "PUT",
    body: JSON.stringify({
      status: "processing",
      transaction_id: chargeId,
      meta_data: [{ key: "_stripe_intent_id", value: intent.id }],
    }),
  });
  await woo(`/orders/${orderId}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Matched to Stripe PaymentIntent ${intent.id} (${intent.payment_method_types}), ` +
            `which confirmed after checkout. Marked processing by the reconciler.`,
    }),
  });
}

export async function run() {
  let fixed = 0;
  for await (const intent of recentSucceededDelayed(LOOKBACK_HOURS)) {
    const order = await resolveOrder(intent);
    const orderId = order ? order.id : intent.metadata.order_id;
    const [action, reason] = decide(order, intent);
    if (action === "orphan") { console.warn(`Intent ${intent.id} has no matching order`); continue; }
    if (action === "skip" || action === "mismatch") {
      if (action === "mismatch") console.warn(`Order ${orderId}: ${reason}`);
      continue;
    }
    console.log(`Order ${orderId}: ${reason}. ${DRY_RUN ? "would fix" : "fixing"}`);
    if (!DRY_RUN) await markProcessing(orderId, intent);
    fixed++;
  }
  console.log(`Done. ${fixed} order(s) ${DRY_RUN ? "to fix" : "fixed"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
